File: curry/NSString+TKCategory.m
//
//  NSStringAddition.m
//  Created by Devin Ross on 10/26/09.
//
/*
 
 tapku || http://github.com/devinross/tapkulibrary
 
 Permission is hereby granted, free of charge, to any person
 obtaining a copy of this software and associated documentation
 files (the "Software"), to deal in the Software without
 restriction, including without limitation the rights to use,
 copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the
 Software is furnished to do so, subject to the following
 conditions:
 
 The above copyright notice and this permission notice shall be
 included in all copies or substantial portions of the Software.
 
 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
 OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
 WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 OTHER DEALINGS IN THE SOFTWARE.
 
 */

#import "NSString+TKCategory.h"
#import <CommonCrypto/CommonDigest.h>

@implementation NSString (TKCategory)


- (NSString*) creditCardType{
	if(self.length < 4) return nil;

	NSString *firstTwo = [self substringToIndex:2];
	NSInteger value = firstTwo.integerValue;
	
	if(value == 34 || value == 37)
		return NSLocalizedString(@"American Express", @"");
	
	else if(value == 36)
		return NSLocalizedString(@"Diners Club", @"");
	
	else if(value == 38)
		return NSLocalizedString(@"Carte Blanche", @"");
	
	else if(value > 50 && value < 56)
		return NSLocalizedString(@"Master Card", @"");
	

	
	
		
	NSString *firstFour = [self substringToIndex:4];
	value = firstFour.integerValue;
	
	
	if(value == 2014 || value == 2149)
		return NSLocalizedString(@"EnRoute", @"");
	
	else if(value == 2131 || value == 1800)
		return NSLocalizedString(@"JCB", @"");
	
	else if(value == 6011)
		return NSLocalizedString(@"Discover", @"");
		
		

	
	NSString *firstThree = [self substringToIndex:3];
	value = firstThree.integerValue;
	
	if(value >= 300 && value < 306) return NSLocalizedString(@"Diners Club", @"");
	
	NSString *firstOne = [self substringToIndex:1];
	value = firstOne.integerValue;
	
	if(value == 3) return NSLocalizedString(@"JCB", @"");
	
	else if(value == 4) return NSLocalizedString(@"Visa", @"");
	

	return NSLocalizedString(@"Unknown", @"");
}


- (BOOL) isValidCreditCardNumber{
	
	NSCharacterSet *nonDecimalsSet = [[NSCharacterSet decimalDigitCharacterSet] invertedSet];
	
	if([self rangeOfCharacterFromSet:nonDecimalsSet].location != NSNotFound)
		return NO;
	
	
	
	
	
	NSInteger luhnChecksum = 0;
	NSInteger subtractionNumber = 48;
	
	for (NSInteger i=0;i<self.length;i++){
		
		NSUInteger count = self.length-1;
		NSInteger doubled = [@([self characterAtIndex:count-i]) integerValue] - subtractionNumber;
		if (i % 2) doubled = doubled*2;
		
		NSString *doubleDigit = [NSString stringWithFormat:@"%@",@(doubled)];
		
		if (doubleDigit.length > 1){
			luhnChecksum += [@([doubleDigit characterAtIndex:0]) integerValue]-subtractionNumber;
			luhnChecksum += [@([doubleDigit characterAtIndex:1]) integerValue]-subtractionNumber;
		}else{
			luhnChecksum += doubled;
		}
	}
	
	
	return luhnChecksum % 10 == 0;
}





- (NSString*) capitalizeSentence{
	if(self.length < 1) return self;
	
	return [self stringByReplacingCharactersInRange:NSMakeRange(0,1) withString:[self substringToIndex:1].capitalizedString];
}

- (BOOL) isEmail{
	
    NSString *emailRegEx =
	@"(?:[a-z0-9!#$%\\&'*+/=?\\^_`{|}~-]+(?:\\.[a-z0-9!#$%\\&'*+/=?\\^_`{|}"
	@"~-]+)*|\"(?:[\\x01-\\x08\\x0b\\x0c\\x0e-\\x1f\\x21\\x23-\\x5b\\x5d-\\"
	@"x7f]|\\\\[\\x01-\\x09\\x0b\\x0c\\x0e-\\x7f])*\")@(?:(?:[a-z0-9](?:[a-"
	@"z0-9-]*[a-z0-9])?\\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?|\\[(?:(?:25[0-5"
	@"]|2[0-4][0-9]|[01]?[0-9][0-9]?)\\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-"  
	@"9][0-9]?|[a-z0-9-]*[a-z0-9]:(?:[\\x01-\\x08\\x0b\\x0c\\x0e-\\x1f\\x21"  
	@"-\\x5a\\x53-\\x7f]|\\\\[\\x01-\\x09\\x0b\\x0c\\x0e-\\x7f])+)\\])"; 
	
    NSPredicate *regExPredicate = [NSPredicate predicateWithFormat:@"SELF MATCHES %@", emailRegEx];  
    return [regExPredicate evaluateWithObject:[self lowercaseString]];  
}

- (NSString*) URLEncode{
	
	NSString *encodedString = (__bridge_transfer NSString *)CFURLCreateStringByAddingPercentEscapes(
																						   NULL,
																						   (__bridge CFStringRef)self,
																						   NULL,
																						   (CFStringRef)@"!*'();:@&=+$,/?%#[]",
																						   kCFStringEncodingUTF8 );
	
	return encodedString;
	
}


- (NSString *) escapeHTML{
	NSMutableString *s = [NSMutableString string];
	
	NSInteger start = 0;
	NSInteger len = [self length];
	NSCharacterSet *chs = [NSCharacterSet characterSetWithCharactersInString:@"<>&\""];
	
	while (start < len) {
		NSRange r = [self rangeOfCharacterFromSet:chs options:0 range:NSMakeRange(start, len-start)];
		if (r.location == NSNotFound) {
			[s appendString:[self substringFromIndex:start]];
			break;
		}
		
		if (start < r.location) {
			[s appendString:[self substringWithRange:NSMakeRange(start, r.location-start)]];
		}
		
		switch ([self characterAtIndex:r.location]) {
			case '<':
				[s appendString:@"&lt;"];
				break;
			case '>':
				[s appendString:@"&gt;"];
				break;
			case '"':
				[s appendString:@"&quot;"];
				break;
				//			case '…':
				//				[s appendString:@"&hellip;"];
				//				break;
			case '&':
				[s appendString:@"&amp;"];
				break;
		}
		
		start = r.location + 1;
	}
	
	return s;
}

- (NSString *) unescapeHTML{
	NSMutableString *s = [NSMutableString string];
	NSMutableString *target = [self mutableCopy];
	NSCharacterSet *chs = [NSCharacterSet characterSetWithCharactersInString:@"&"];
	
	while ([target length] > 0) {
		NSRange r = [target rangeOfCharacterFromSet:chs];
		if (r.location == NSNotFound) {
			[s appendString:target];
			break;
		}
		
		if (r.location > 0) {
			[s appendString:[target substringToIndex:r.location]];
			[target deleteCharactersInRange:NSMakeRange(0, r.location)];
		}
		
		if ([target hasPrefix:@"&lt;"]) {
			[s appendString:@"<"];
			[target deleteCharactersInRange:NSMakeRange(0, 4)];
		} else if ([target hasPrefix:@"&gt;"]) {
			[s appendString:@">"];
			[target deleteCharactersInRange:NSMakeRange(0, 4)];
		} else if ([target hasPrefix:@"&quot;"]) {
			[s appendString:@"\""];
			[target deleteCharactersInRange:NSMakeRange(0, 6)];
		} else if ([target hasPrefix:@"&#39;"]) {
			[s appendString:@"'"];
			[target deleteCharactersInRange:NSMakeRange(0, 5)];
		} else if ([target hasPrefix:@"&amp;"]) {
			[s appendString:@"&"];
			[target deleteCharactersInRange:NSMakeRange(0, 5)];
		} else if ([target hasPrefix:@"&hellip;"]) {
			[s appendString:@"…"];
			[target deleteCharactersInRange:NSMakeRange(0, 8)];
		} else {
			[s appendString:@"&"];
			[target deleteCharactersInRange:NSMakeRange(0, 1)];
		}
	}
	return s;
}

- (NSString*) stringByRemovingHTML{
	
	NSString *html = self;
    NSScanner *thescanner = [NSScanner scannerWithString:html];
    NSString *text = nil;
    while ([thescanner isAtEnd] == NO) {
		[thescanner scanUpToString:@"<" intoString:NULL];
		[thescanner scanUpToString:@">" intoString:&text];
		html = [html stringByReplacingOccurrencesOfString:[NSString stringWithFormat:@"%@>",text] withString:@" "];
    }
	return html;
}

- (NSString *) md5sum{
	unsigned char digest[CC_MD5_DIGEST_LENGTH], i;
	CC_MD5([self UTF8String], (uint32_t)[self lengthOfBytesUsingEncoding:NSUTF8StringEncoding], digest);
	NSMutableString *ms = [NSMutableString string];
	for (i=0;i<CC_MD5_DIGEST_LENGTH;i++) {
		[ms appendFormat: @"%02x", (int)(digest[i])];
	}
	return [ms copy];
}

- (BOOL) hasString:(NSString*)substring{
	return !([self rangeOfString:substring].location == NSNotFound);
}





- (NSString*) formattedPhoneNumberWithLastCharacterRemoved:(BOOL)deleteLastChar{
	if(self.length<1) return @"";
	
	NSError *error = NULL;
	NSRegularExpression *regex = [NSRegularExpression regularExpressionWithPattern:@"[\\s-\\(\\)]" options:NSRegularExpressionCaseInsensitive error:&error];
	NSString *digits = [regex stringByReplacingMatchesInString:self options:0 range:NSMakeRange(0, [self length]) withTemplate:@""];
	
	
	// should we delete the last digit?
	if(deleteLastChar) 
		digits = [digits substringToIndex:digits.length - 1];

	
	// 123 456 7890
	// format the number.. if it's less then 7 digits.. then use this regex.
	BOOL leadingOne = [digits hasPrefix:@"1"];

	if((digits.length > 11 && leadingOne) || (digits.length > 10 && !leadingOne))
		return digits;
	
	NSStringCompareOptions opt = NSRegularExpressionSearch;
	NSRange range = NSMakeRange(0, digits.length);
	
	NSString *occurence, *replace;

	
	if(digits.length < 5 && leadingOne){
		occurence = @"(\\d{1})(\\d+)";
		replace = @"$1 ($2)";
		
	}else if(digits.length < 8 && leadingOne){
		occurence = @"(\\d{1})(\\d{3})(\\d+)";
		replace = @"$1 ($2) $3";
		
	}else if(digits.length<7){
		occurence = @"(\\d{3})(\\d+)";
		replace = @"($1) $2";
		
	}else if(digits.length > 6 && leadingOne){
		occurence = @"(\\d{1})(\\d{3})(\\d{3})(\\d+)";
		replace = @"$1 ($2) $3-$4";
		
	}else{
		occurence = @"(\\d{3})(\\d{3})(\\d+)";
		replace = @"($1) $2-$3";
	}

	digits = [digits stringByReplacingOccurrencesOfString:occurence withString:replace options:opt range:range];

	return digits;
}


@end

File: CurrySwift/NSString+TKCategory.swift
//
//  NSString+TKCategory.swift
//  Created by Devin Ross on 10/26/09.
//
/*

curry || https://github.com/devinross/curry

Permission is hereby granted, free of charge, to any person
obtaining a copy of this software and associated documentation
files (the "Software"), to deal in the Software without
restriction, including without limitation the rights to use,
copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the
Software is furnished to do so, subject to the following
conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.

*/

import Foundation
import CryptoKit


/** Additional functionality for `NSString`. */
extension NSString {
	
	/** Returns `YES` if a string is a valid email address, otherwise `NO`.
	@return True if the string is formatted properly as an email address.
	*/
	@objc(isEmail)
	public var isEmail: Bool {
		
		let badCharacters = CharacterSet(charactersIn: " !\\#$%(),:;<>[]/`|")
		if rangeOfCharacter(from: badCharacters).location != NSNotFound {
			return false
		}
		
		let emailRegEx =
			"(?:[a-z0-9!#$%\\&'*+/=?\\^_`{|}~-]+(?:\\.[a-z0-9!#$%\\&'*+/=?\\^_`{|}" +
			"~-]+)*|\"(?:[\\x01-\\x08\\x0b\\x0c\\x0e-\\x1f\\x21\\x23-\\x5b\\x5d-\\" +
			"x7f]|\\\\[\\x01-\\x09\\x0b\\x0c\\x0e-\\x7f])*\")@(?:(?:[a-z0-9](?:[a-" +
			"z0-9-]*[a-z0-9])?\\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?|\\[(?:(?:25[0-5" +
			"]|2[0-4][0-9]|[01]?[0-9][0-9]?)\\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-" +
			"9][0-9]?|[a-z0-9-]*[a-z0-9]:(?:[\\x01-\\x08\\x0b\\x0c\\x0e-\\x1f\\x21" +
			"-\\x5a\\x53-\\x7f]|\\\\[\\x01-\\x09\\x0b\\x0c\\x0e-\\x7f])+)\\])"
		
		let predicate = NSPredicate(format: "SELF MATCHES %@", emailRegEx)
		return predicate.evaluate(with: lowercased)
	}
	
	/** Returns a `NSString` that removes blank space from the start and end of string. */
	@objc public var stripWhitespace: String {
		trimmingCharacters(in: .whitespaces)
	}
	
	/** Returns a `NSString` that removes blank space and new lines from the start and end of string. */
	@objc public var stripWhitepspaceAndNewlines: String {
		trimmingCharacters(in: .whitespacesAndNewlines)
	}
	
	/** Returns a `NSString` that is URL friendly. */
	@objc public var URLEncode: String {
		var allowed = CharacterSet.urlQueryAllowed
		allowed.remove(charactersIn: ";?/:#& =+$,%<>~%")
		return addingPercentEncoding(withAllowedCharacters: allowed) ?? ""
	}
	
	/** Returns a `NSString` that properly replaces HTML specific character sequences. */
	@objc public var escapeHTML: String {
		
		let escaped = NSMutableString()
		let specials = CharacterSet(charactersIn: "<>&\"")
		let total = length
		var start = 0
		
		while start < total {
			
			let searchRange = NSRange(location: start, length: total - start)
			let found = rangeOfCharacter(from: specials, options: [], range: searchRange)
			
			guard found.location != NSNotFound else {
				escaped.append(substring(from: start))
				break
			}
			
			if start < found.location {
				escaped.append(substring(with: NSRange(location: start, length: found.location - start)))
			}
			
			switch character(at: found.location) {
				case 0x3C: escaped.append("&lt;")   // <
				case 0x3E: escaped.append("&gt;")   // >
				case 0x22: escaped.append("&quot;") // "
				case 0x26: escaped.append("&amp;")  // &
				default: break
			}
			
			start = found.location + 1
		}
		
		return escaped as String
	}
	
	/** Returns a `NSString` that properly formats text for HTML. */
	@objc public var unescapeHTML: String {
		
		let unescaped = NSMutableString()
		let target = NSMutableString(string: self as String)
		let ampersand = CharacterSet(charactersIn: "&")
		
		let entities: [(entity: String, replacement: String)] = [
			("&lt;", "<"),
			("&gt;", ">"),
			("&quot;", "\""),
			("&#39;", "'"),
			("&amp;", "&"),
			("&hellip;", "\u{2026}"),
		]
		
		while target.length > 0 {
			
			let found = target.rangeOfCharacter(from: ampersand)
			
			guard found.location != NSNotFound else {
				unescaped.append(target as String)
				break
			}
			
			if found.location > 0 {
				unescaped.append(target.substring(to: found.location))
				target.deleteCharacters(in: NSRange(location: 0, length: found.location))
			}
			
			if let match = entities.first(where: { target.hasPrefix($0.entity) }) {
				unescaped.append(match.replacement)
				target.deleteCharacters(in: NSRange(location: 0, length: (match.entity as NSString).length))
			} else {
				unescaped.append("&")
				target.deleteCharacters(in: NSRange(location: 0, length: 1))
			}
		}
		
		return unescaped as String
	}
	
	/** Returns a `NSString` that removes HTML elements. */
	@objc public var stringByRemovingHTML: String {
		
		var html = self as String
		let scanner = Scanner(string: html)
		
		while !scanner.isAtEnd {
			_ = scanner.scanUpToString("<")
			guard let text = scanner.scanUpToString(">") else { break }
			html = html.replacingOccurrences(of: "\(text)>", with: " ")
		}
		
		return html
	}
	
	/** Returns an MD5 string of from the given `NSString`. */
	@objc public var md5sum: String {
		Insecure.MD5
			.hash(data: Data((self as String).utf8))
			.map { String(format: "%02x", $0) }
			.joined()
	}
	
	/** Returns `YES` is a string has the substring, otherwise `NO`.
	@param substring The substring.
	@return `YES` if the substring is contained in the string, otherwise `NO`.
	*/
	@objc(hasString:)
	public func hasString(_ substring: String) -> Bool {
		range(of: substring).location != NSNotFound
	}
	
	/** Returns the length of the string minus the whitespace characters. */
	@objc public var lengthWithoutWhitespace: Int {
		(stringWithoutWhitespace as NSString).length
	}
	
	/** Returns a string without the whitespace characters. */
	@objc public var stringWithoutWhitespace: String {
		components(separatedBy: .whitespacesAndNewlines).joined()
	}
	
	/* Returns a string with the first word in each sentence as capitalized. */
	@objc public var capitalizeSentence: String {
		guard length > 0 else { return self as String }
		return replacingCharacters(in: NSRange(location: 0, length: 1),
		                           with: substring(to: 1).capitalized)
	}
	
	/** Returns the credit card type based on the first four digits of the card number.
	@return A credit card company name if the number matches a company otherwise nil.
	*/
	@objc public var creditCardType: String? {
		
		guard length >= 4 else { return nil }
		
		switch (substring(to: 2) as NSString).integerValue {
			case 34, 37:  return NSLocalizedString("American Express", comment: "")
			case 36:      return NSLocalizedString("Diners Club", comment: "")
			case 38:      return NSLocalizedString("Carte Blanche", comment: "")
			case 51...55: return NSLocalizedString("Master Card", comment: "")
			default: break
		}
		
		switch (substring(to: 4) as NSString).integerValue {
			case 2014, 2149: return NSLocalizedString("EnRoute", comment: "")
			case 2131, 1800: return NSLocalizedString("JCB", comment: "")
			case 6011:       return NSLocalizedString("Discover", comment: "")
			default: break
		}
		
		if case 300...305 = (substring(to: 3) as NSString).integerValue {
			return NSLocalizedString("Diners Club", comment: "")
		}
		
		switch (substring(to: 1) as NSString).integerValue {
			case 3: return NSLocalizedString("JCB", comment: "")
			case 4: return NSLocalizedString("Visa", comment: "")
			default: break
		}
		
		return NSLocalizedString("Unknown", comment: "")
	}
	
	/** Returns YES if the card number passes the Luhn algorithm. No spaces in the card number.
	@return Returns YES if the card is a valid credit card number, otherwise NO.
	*/
	@objc(isValidCreditCardNumber)
	public var isValidCreditCardNumber: Bool {
		
		guard rangeOfCharacter(from: CharacterSet.decimalDigits.inverted).location == NSNotFound else {
			return false
		}
		
		var checksum = 0
		
		for (offset, character) in (self as String).reversed().enumerated() {
			guard let digit = character.wholeNumberValue else { return false }
			// Every second digit from the right is doubled, and a two digit
			// result contributes each of its digits separately.
			let value = offset.isMultiple(of: 2) ? digit : digit * 2
			checksum += value > 9 ? (value / 10) + (value % 10) : value
		}
		
		return checksum % 10 == 0
	}
	
	/** Returns a formatted phone number. So it'd transform a string 18005555555 would be 1-800-555-5555.
	@param deleteLastChar Remove the last character from the string.
	@return Returns an `NSString` that transform a string of numbers into a string human readable phone number.
	*/
	@objc(formattedPhoneNumberWithLastCharacterRemoved:)
	public func formattedPhoneNumber(withLastCharacterRemoved deleteLastChar: Bool) -> String {
		formattedPhoneNumber(withLastCharacterRemoved: deleteLastChar, limit: 15)
	}
	
	/** Returns a formatted phone number, capping an over-long number at `limit` digits.
	@param deleteLastChar Remove the last character from the string.
	@param limit The most characters returned when the number is too long to format.
	*/
	@objc(formattedPhoneNumberWithLastCharacterRemoved:limit:)
	public func formattedPhoneNumber(withLastCharacterRemoved deleteLastChar: Bool, limit: Int) -> String {
		
		guard length > 0 else { return "" }
		
		var digits = (self as String).replacingOccurrences(of: "[\\s-\\(\\)]",
		                                                   with: "",
		                                                   options: [.regularExpression, .caseInsensitive])
		
		// should we delete the last digit?
		if deleteLastChar {
			guard !digits.isEmpty else { return "" }
			digits.removeLast()
		}
		
		// 123 456 7890
		// format the number.. if it's less then 7 digits.. then use this regex.
		let count = (digits as NSString).length
		let leadingOne = digits.hasPrefix("1")
		
		if (count > 11 && leadingOne) || (count > 10 && !leadingOne) {
			return (digits as NSString).substring(to: min(limit, count))
		}
		
		let occurrence: String, replacement: String
		
		if count < 5 && leadingOne {
			occurrence = "(\\d{1})(\\d+)"
			replacement = "$1 ($2)"
		} else if count < 8 && leadingOne {
			occurrence = "(\\d{1})(\\d{3})(\\d+)"
			replacement = "$1 ($2) $3"
		} else if count < 7 {
			occurrence = "(\\d{3})(\\d+)"
			replacement = "($1) $2"
		} else if count > 6 && leadingOne {
			occurrence = "(\\d{1})(\\d{3})(\\d{3})(\\d+)"
			replacement = "$1 ($2) $3-$4"
		} else {
			occurrence = "(\\d{3})(\\d{3})(\\d+)"
			replacement = "($1) $2-$3"
		}
		
		return digits.replacingOccurrences(of: occurrence,
		                                   with: replacement,
		                                   options: .regularExpression,
		                                   range: digits.startIndex ..< digits.endIndex)
	}
	
}
